docs/user/tutorial-asgi.rst

```diff
@@ -0,0 +1,1090 @@
+
+        def list_images(self):
+            return sorted(self._images.values(), key=lambda item: item.modified)
+
+        async def save(self, image_id, data):
+            loop = asyncio.get_running_loop()
+            image = await loop.run_in_executor(None, self._load_from_bytes, data)
+            converted = await loop.run_in_executor(None, self._convert, image)
+
+            path = os.path.join(self.config.storage_path, image_id)
+            async with aiofiles.open(path, 'wb') as output:
+                await output.write(converted)
+
+            stored = Image(self.config, image_id, image.size)
+            self._images[image_id] = stored
+            return stored
+
+Here we store data using ``aiofiles``, and run ``Pillow`` image transformation
+functions in a threadpool executor, hoping that at least some of them release
+the GIL during processing.
+
+Images Resource(s)
+------------------
+
+In the ASGI flavor of Falcon, all responder methods, hooks and middleware
+methods must be awaitable coroutines. With that in mind, let's go on to
+implement the image collection, and the individual image resources (the code
+below should go into ``images.py``):
+
+.. code:: python
+
+    import aiofiles
+    import falcon
+
+
+    class Images:
+
+        def __init__(self, config, store):
+            self.config = config
+            self.store = store
+
+        async def on_get(self, req, resp):
+            resp.media = [image.serialize() for image in self.store.list_images()]
+
+        async def on_get_image(self, req, resp, image_id):
+            image = self.store.get(str(image_id))
+            resp.stream = await aiofiles.open(image.path, 'rb')
+            resp.content_type = falcon.MEDIA_JPEG
+
+        async def on_post(self, req, resp):
+            data = await req.stream.read()
+            image_id = str(self.config.uuid_generator())
+            image = await self.store.save(image_id, data)
+
+            resp.location = image.uri
+            resp.media = image.serialize()
+            resp.status = falcon.HTTP_201
+
+Since the first iteration on the ``Images`` class is quite lean, we opted for
+implementing two resources, image collection (which supports ``GET`` for
+listing the collection, and ``POST`` for uploading a new image) and single
+image (which supports ``GET`` for downloading the image), under one class
+employing responder name suffixes.
+
+If the application continues to grow in complexity, it might get worth to make
+the code cleaner by splitting classes to strictly represent one RESTful
+resource per class. See also: :ref:`recommended-route-layout`
+
+.. note::
+   Here, we serve the image by simply assigning an open ``aiofiles`` file to
+   :attr:`resp.stream <falcon.asgi.Response.stream>`.
+
+.. warning::
+   In production deployment, serving files directly from the web server, rather
+   than through the Falcon ASGI app, will likely be more efficient, and therefore
```

Comment:
I think we can safely remove the `likely` here :)

Reply:
I'd prefer to leave it though. There are slow web servers out there written in other languages too, not only Python.

Reply:
ok